## <span style='color:Green'> First, I installed some libraries in order to proceed with the scraping. </span>

In [156]:
pip install lxml html5lib beautifulsoup4

Note: you may need to restart the kernel to use updated packages.


In [157]:
import pandas as pd

In [158]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [159]:
df = pd.read_html(url)

## <span style='color:Green'> The dataframe from the Wikipedia page: </span>

In [160]:
df[0].head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [161]:
df = df[0]

## <span style='color:Green'> In the code below, the cells with a Borough that is 'Not assigned' are ignored. </span>

In [162]:
(df == 'Not assigned').any()

Postal Code      False
Borough           True
Neighbourhood     True
dtype: bool

In [163]:
import numpy as np
df.loc[:, 'Borough'].replace('Not assigned', np.nan, inplace=True)

In [164]:
df.head()

Postal Code           Borough              Neighbourhood
0         M1A               NaN               Not assigned
1         M2A               NaN               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [165]:
df.dropna(subset=['Borough'], axis=0, inplace=True)

In [166]:
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [167]:
df.reset_index(drop=True)

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## <span style='color:Green'> More than one Neighborhood can exist in one postal code area; so, let us group by Postal Code: </span>

In [168]:
df.groupby('Postal Code')

In [169]:
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

## <span style='color:Green'> In the cell below, I show that there are no cell in Neighborhood columns with 'Not assigned' value, so we can skip the replace of values. </span>

In [170]:
(df == 'Not assigned').any()

Postal Code      False
Borough          False
Neighbourhood    False
dtype: bool

## <span style='color:Green'> In conclusion, the dataframe has 103 rows and 3 columns, as shown below: </span>

In [171]:
df.shape

(103, 3)

# <span style='color:Blue'> -> The second assignment begins from this point. </span>

## <span style='color:Blue'> I used the .csv file because I could not install nor import 'geocoder'. </span>

In [172]:
import pandas as pd
url = 'https://cocl.us/Geospatial_data'
geo_data = pd.read_csv(url)

In [173]:
geo_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## <span style='color:Blue'> So, we can do the join between the two dataframes, on the common key 'Postal Code': </span>

In [174]:
df = df.join(geo_data.set_index('Postal Code'), on='Postal Code')

In [175]:
df = df.reset_index(drop=True)

In [176]:
df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

# <span style='color:Red'> -> The third assignment begins from this point. </span>

### <span style='color:Red'> Here I created the API GET request for the URL:  </span>

In [177]:
CLIENT_ID = 'GPAXIIUH2BT4LFICUWK2QHG2MVDRXVVPCKJZJ24EOWGNSDFB'
CLIENT_SECRET = 'F4QDIQXCZOYTTJ5W4I5FO04N0IEAMORNTAW05LPXYFC2QSXC'
VERSION = '20180605'
neighborhood_latitude = df.loc[0, 'Latitude']
neighborhood_longitude = df.loc[0, 'Longitude']
radius = 500
LIMIT = 100

In [178]:
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url

'https://api.foursquare.com/v2/venues/explore?&client_id=GPAXIIUH2BT4LFICUWK2QHG2MVDRXVVPCKJZJ24EOWGNSDFB&client_secret=F4QDIQXCZOYTTJ5W4I5FO04N0IEAMORNTAW05LPXYFC2QSXC&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [179]:
import json
import requests
from pandas.io.json import json_normalize
results = requests.get(url).json()

### <span style='color:Red'> Here I used the function that extract the category of the venue, defined in the lab of the course: </span>

In [180]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [181]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

In [182]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare.


### <span style='color:Red'> Here I used the function that give us venues near some others, defined in the lab of the course: </span>

In [183]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [184]:
toronto_venues = getNearbyVenues(names=df['Neighbourhood'], 
                                   latitudes=df['Latitude'], 
                                   longitudes=df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [185]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

In [186]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                4   
Woodbine Heights                                                      7   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              5   
Alderwood, Long Branch                                                 7   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     22   
...                                                                  ...   
Willowdale, Willowdale West                                            5   
Woburn                                                                 4   
Woodbine Heights                                                       7   
York Mills West                                                        2   
York Mills, Silver Hills                                               1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            5               5   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               4               4   
Woodbine Heights                                     7               7   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      5   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         4   
Woodbine Heights                                               7   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     5  
Alderwood, Long Branch                

In [187]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 273 uniques categories.


### <span style='color:Red'> Let us create a dataframe where we can read the different types of venues in each neighbourhood. </span>

In [188]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2  Victoria Village                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                   0                              0                 0   
1                   0                              0                 0   
2                   0                              0                 0   
3                   0                              0                 0   
4                   0                              0                 0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 274 columns]

In [189]:
toronto_onehot.shape

(2136, 274)

In [190]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
91                      Willowdale, Willowdale West                0.0   
92                                           Woburn                0.0   
93                                 Woodbine Heights                0.0   
94                                  York Mills West                0.0   
95                         York Mills, Silver Hills                0.0   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                 0.0      0.0                 0.0           0.0   
1                 0.0      0.0                 0.0           0.0   
2                 0.0      0.0                 0.0           0.0   
3                 0.0      0.0                 0.0           0.0   
4                 0.0      0.0                 0.0           0.0   
..                ...      ...                 ...           ...   
91                0.0      0.0                 0.0           0.0   
92                0.0      0.0                 0.0           0.0   
93                0.0      0.0                 0.0           0.0   
94                0.0      0.0                 0.0           0.0   
95                0.0      0.0                 0.0           0.0   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0              0.0              0.0               0.0             0.000000   
1              0.0              0.0               0.0             0.000000   
2              0.0              0.0               0.0             0.000000   
3              0.0              0.0               0.0             0.000000   
4              0.0              0.0               0.0             0.045455   
..             ...              ...               ...                  ...   
91             0.0              0.0               0.0             0.000000   
92             0.0              0.0               0.0             0.000000   
93             0.0              0.0               0.0             0.000000   
94             0.0              0.0               0.0             0.000000   
95             0.0              0.0               0.0             0.000000   

    ...  Train Station  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
0   ...            0.0                 0.0                            0.0   
1   ...            0.0                 0.0                            0.0   
2   ...            0.0                 0.0                            0.0   
3   ...            0.0                 0.0                            0.0   
4   ...            0.0                 0.0                            0.0   
..  ...            ...                 ...                            ...   
91  ...            0.0                 0.0                            0.0   
92  ...            0.0                 0.0                            0.0   
93  ...            0.0                 0.0                            0.0   
94  ...            0.0                 0.0                            0.0   
95  ...            0.0                 0.0                            0.0   

    Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0                0.0                    0.0              0.0       0.0   
1                0.0                    0.0              0.0       0.0   
2                0.0                    0.0              0.0       0.0   
3                0.0                    0.0              0.0       0.0   
4                0.0                    0.0              0.0       0.0   
..               ...      

In [191]:
toronto_grouped.shape

(96, 274)

In [192]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Clothing Store   0.2
1             Breakfast Spot   0.2
2                     Lounge   0.2
3               Skating Rink   0.2
4  Latin American Restaurant   0.2


----Alderwood, Long Branch----
         venue  freq
0  Pizza Place  0.29
1          Gym  0.14
2  Coffee Shop  0.14
3          Pub  0.14
4     Pharmacy  0.14


----Bathurst Manor, Wilson Heights, Downsview North----
         venue  freq
0         Bank  0.10
1  Coffee Shop  0.10
2     Pharmacy  0.05
3        Diner  0.05
4  Gas Station  0.05


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1   Chinese Restaurant  0.25
2                 Bank  0.25
3                 Café  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0      Sandwich Place  0.09
1         Coffee Shop  0.09
2  Italian Restaurant  0.09
3            Pharmacy  0.05
4       Grocery Store  0.05


----Berczy Park

                 venue  freq
0           Playground   0.5
1                Trail   0.5
2   Miscellaneous Shop   0.0
3  Moroccan Restaurant   0.0
4  Monument / Landmark   0.0


----New Toronto, Mimico South, Humber Bay Shores----
                 venue  freq
0                 Café  0.14
1          Coffee Shop  0.14
2          Pizza Place  0.07
3  American Restaurant  0.07
4           Restaurant  0.07


----North Park, Maple Leaf Park, Upwood Park----
                        venue  freq
0                      Bakery  0.33
1  Construction & Landscaping  0.33
2                        Park  0.33
3           Accessories Store  0.00
4          Miscellaneous Shop  0.00


----North Toronto West, Lawrence Park----
                  venue  freq
0           Coffee Shop  0.11
1        Clothing Store  0.11
2           Yoga Studio  0.06
3  Fast Food Restaurant  0.06
4                   Spa  0.06


----Northwest, West Humber - Clairville----
                 venue  freq
0        Garden Center  0.25
1 

                venue  freq
0        Skating Rink  0.29
1                Park  0.14
2  Athletics & Sports  0.14
3          Beer Store  0.14
4         Curling Ice  0.14


----York Mills West----
                       venue  freq
0          Convenience Store   0.5
1                       Park   0.5
2          Accessories Store   0.0
3  Middle Eastern Restaurant   0.0
4        Moroccan Restaurant   0.0


----York Mills, Silver Hills----
                 venue  freq
0  Martial Arts School   1.0
1   Miscellaneous Shop   0.0
2                Motel   0.0
3  Moroccan Restaurant   0.0
4  Monument / Landmark   0.0




In [193]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [194]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt        Clothing Store   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Breakfast Spot          Skating Rink   
1        Sandwich Place           Coffee Shop                   Pub   
2                  Bank              Pharmacy        Ice Cream Shop   
3                  Café    Chinese Restaurant                  Bank   
4    Italian Restaurant           Coffee Shop      Greek Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Latin American Restaurant     Electronics Store      Doner Restaurant   
1                   Pharmacy                   Gym      Greek Restaurant   
2                Bridal Shop         Shopping Mall        Sandwich Place   
3        Distribution Center               Dog Run      Doner Restaurant   
4            Thai Restaurant             Locksmith          Liquor Store   

     8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0               Donut Shop                  Drugstore    Dumpling Restaurant  
1           Discount Store           Department Store           Dessert Shop  
2                    Diner  Middle Eastern Restaurant             Restaurant  
3               Donut Shop                  Drugstore            Yoga Studio  
4  Comfort Food Restaurant                  Juice Bar                Butcher

In [195]:
from sklearn.cluster import KMeans
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [196]:
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df

toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio    Dumpling Restaurant   
1          Hockey Arena           Coffee Shop  Portuguese Restaurant   
2                Bakery                   Pub                   Park   
3     Accessories Store              Boutique              Gift Shop   
4           Yoga Studio    College Auditorium                    Bar   

    5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
0     Distribution Center                     Dog Run      Doner Restaurant   
1            Intersection  Construction & Landscaping    Dim Sum Restaurant   
2          Breakfast Spot                        Café               Theater   
3  Furniture / Home Store                 Event Space           Coffee Shop   
4                Beer Bar               Smoothie Shop        Sandwich Place   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0            Donut Shop              Drugstore      Electronics Store  
1           Event Space   Ethiopian Restaurant            Escape Room  
2                   Spa                Brewery             Shoe Store  
3         Women's Store  Vietnamese Restaurant        Airport Service  
4                  Café             Restaurant     Chinese Restaurant

In [197]:
toronto_merged[['Cluster Labels']]

Cluster Labels
0               0.0
1               1.0
2               1.0
3               1.0
4               1.0
..              ...
98              1.0
99              1.0
100             1.0
101             1.0
102             1.0

[103 rows x 1 columns]

### <span style='color:Red'> Let us drop the NaN values in Cluster Labels: </span>

In [198]:
toronto_merged['Cluster Labels']=toronto_merged['Cluster Labels'].fillna(0).astype('int')

In [199]:
toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               0                  Park   
1  43.725882 -79.315572               1           Pizza Place   
2  43.654260 -79.360636               1           Coffee Shop   
3  43.718518 -79.464763               1        Clothing Store   
4  43.662301 -79.389494               1           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio    Dumpling Restaurant   
1          Hockey Arena           Coffee Shop  Portuguese Restaurant   
2                Bakery                   Pub                   Park   
3     Accessories Store              Boutique              Gift Shop   
4           Yoga Studio    College Auditorium                    Bar   

    5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
0     Distribution Center                     Dog Run      Doner Restaurant   
1            Intersection  Construction & Landscaping    Dim Sum Restaurant   
2          Breakfast Spot                        Café               Theater   
3  Furniture / Home Store                 Event Space           Coffee Shop   
4                Beer Bar               Smoothie Shop        Sandwich Place   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0            Donut Shop              Drugstore      Electronics Store  
1           Event Space   Ethiopian Restaurant            Escape Room  
2                   Spa                Brewery             Shoe Store  
3         Women's Store  Vietnamese Restaurant        Airport Service  
4                  Café             Restaurant     Chinese Restaurant

In [200]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [201]:
!pip install folium

In [202]:
import folium

### <span style='color:Red'> So, let's visualize the resulting clusters: </span>

In [203]:
map_clusters = folium.Map(location=[43.651070, -79.347015], zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters